core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java

```diff
@@ -502,7 +509,7 @@ assert getByAllocationId(failedShard.shardId(), failedShard.allocationId().getId
                         UnassignedInfo primaryFailedUnassignedInfo = new UnassignedInfo(UnassignedInfo.Reason.PRIMARY_FAILED,
                             "primary failed while replica initializing", null, 0, unassignedInfo.getUnassignedTimeInNanos(),
                             unassignedInfo.getUnassignedTimeInMillis(), false, AllocationStatus.NO_ATTEMPT);
-                        failShard(logger, replicaShard, primaryFailedUnassignedInfo, indexMetaData);
+                        failShard(logger, replicaShard, primaryFailedUnassignedInfo, indexMetaData, routingChanges);
                     }
                 }
             }
```

Comment:
hehe, good catch. I initially chose not to record those changes because they looked boring (and they are not directly relevant for the "isChanged" flag, because removal of "failedShard" will mark the flag to changed anyhow). On second thought (and for sake of completeness), it makes sense to add them as well to the RoutingChanges API. I've pushed aeff8f1 to address this.